src/lib/badge/badge.ts

```diff
@@ -0,0 +1,221 @@
+    this._position = val;
+    this._isAbove = val.indexOf('below') === -1;
+    this._isAfter = val.indexOf('before') === -1;
+  }
+  get matBadgePosition(): string {
+    return this._position;
+  }
+  private _position: string = 'above after';
+
+  _id: number = nextId++;
+  _isAbove: boolean = true;
+  _isAfter: boolean = true;
+
+  constructor(public _renderer: Renderer2, public _elementRef: ElementRef) {}
+
+  /** Clears the previous badge content */
+  _clearContents(): void {
+    const contents = document.getElementById(`badge-content-${this._id}`);
+    if (contents) {
+      this._renderer.removeChild(this._elementRef.nativeElement, contents);
+    }
+  }
+}
+
+/** Directive to display a text badge. */
+@Directive({
+  selector: '[matBadge]',
+  inputs: ['matBadgeColor', 'matBadgeOverlap', 'matBadgePosition'],
+  host: {
+    'class': 'mat-badge',
+    '[class.mat-badge-overlap]': '_overlap',
+    '[class.mat-badge-above]': '_isAbove',
+    '[class.mat-badge-below]': '!_isAbove',
+    '[class.mat-badge-before]': '!_isAfter',
+    '[class.mat-badge-after]': '_isAfter',
+  },
+})
+export class MatBadge extends MatBadgeBase {
+
+  @Input('matBadge')
+  set content(val: string) {
+    this._content = val;
+    this._clearContents();
+    this._setContent();
+  }
+  get content(): string { return this._content; }
+  private _content: string;
+
+  constructor(_renderer: Renderer2, _elementRef: ElementRef) {
+    super(_renderer, _elementRef);
+  }
+
+  /** Injects a span element into the DOM with the content. */
+  private _setContent(): void {
+    const content = document.createElement('span');
+    content.setAttribute('id', `badge-content-${this._id}`);
+    content.classList.add('mat-badge-text');
+    content.innerText = this.content;
+    this._elementRef.nativeElement.appendChild(content);
+  }
+}
+
+/** Directive to display a font icon badge. */
+@Directive({
+  selector: '[matIconBadge]',
+  inputs: ['matBadgeColor', 'matBadgeOverlap', 'matBadgePosition'],
+  host: {
+    'class': 'mat-badge',
+    '[class.mat-badge-overlap]': '_overlap',
+    '[class.mat-badge-above]': '_isAbove',
+    '[class.mat-badge-below]': '!_isAbove',
+    '[class.mat-badge-before]': '!_isAfter',
+    '[class.mat-badge-after]': '_isAfter',
+  },
+})
+export class MatIconBadge extends MatBadgeBase {
+
+  @Input('matIconBadge')
+  set icon(val: string) {
+    this._icon = val;
+    this._clearContents();
+    this._setFontIcon();
+  }
+  get icon(): string { return this._icon; }
+  private _icon: string;
+
+  /** Font set to use for the icon */
+  @Input() matBadgeFontSet: string;
+
+  constructor(
+      private _iconRegistry: MatIconRegistry,
+      _elementRef: ElementRef,
+      _renderer: Renderer2) {
+    super(_renderer, _elementRef);
+  }
+
+  /**  Gets the font icon set to use. */
+  private _getFontSet(): string {
+    return this.matBadgeFontSet ?
+        this._iconRegistry.classNameForFontAlias(this.matBadgeFontSet) :
```

Comment:
This assumes that the consumer will use Material icons, which may not be the case. Why not let them provide the icon element themselves?

Reply:
This was built to implement Material Icons. I think custom icons might be out of scope for this. @jelbourn - thoughts?

Although I'm curious how you would implement providing the icon element themselves, could you elaborate?

Reply:
If you convert it into a component and allow content projection, users will be able to override it as they want. Another approach can be to provide an `ng-template`, but that might be overkill for this case (see https://github.com/angular/material2/pull/7482).